### PR TITLE
[BUG] Prevent Duplicated Benchmarking Results

### LINKED ISSUE #8569: [BUG] benchmark.run() seems to be printing output from previous ipynb runs even after restarting the kernel
**Describe the bug**
It seems that output from previous runs of benchmark.run() is shown after restarting the kernel in a Jupyter Notebook.

To replicate this behaviour run the code below and change the estimator_id of a model. Restart the kernel and rerun the notebook.

```python
from sktime.benchmarking.forecasting import ForecastingBenchmark
from sktime.performance_metrics.forecasting import MeanAbsolutePercentageError
from sktime.split import SingleWindowSplitter
from sktime.forecasting.ttm import TinyTimeMixerForecaster
from sktime.forecasting.moirai_forecaster import MOIRAIForecaster

from sktime.utils._testing.hierarchical import _make_hierarchical

y = _make_hierarchical(n_columns=2, hierarchy_levels=(2, 3), max_timepoints=50, min_timepoints=50)

benchmark = ForecastingBenchmark()
scorers = [MeanAbsolutePercentageError()]
single_fold_splitter = SingleWindowSplitter(window_length = 16, fh=range(1,4)) # 0 would include the last value of the in-sample data (!)
benchmark.add_task(
        y,
        single_fold_splitter,
        scorers
    )
benchmark.add_estimator(
   TinyTimeMixerForecaster(),
   estimator_id="TTM"
)

benchmark.add_estimator(
   MOIRAIForecaster(checkpoint_path=f"sktime/moirai-1.0-R-small"), 
   estimator_id="MORAI_SMALL"
)

benchmark.add_estimator(
   MOIRAIForecaster(checkpoint_path=f"sktime/moirai-1.0-R-large"), 
   estimator_id="MORAI_LARGE"
)

results = benchmark.run("output/benchmark_results.csv", force_rerun=True)

stat_results = results[["model_id", "MeanAbsolutePercentageError_mean", "runtime_secs"]]
stat_results = stat_results.rename(columns={ "MeanAbsolutePercentageError_mean": "MAPE"}).sort_values(by="MAPE")
stat_results["runtime_secs"] = stat_results["runtime_secs"].round(3)
stat_results
```

Cell output:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>model_id</th>
      <th>MAPE</th>
      <th>runtime_secs</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>8</th>
      <td>moirai_large</td>
      <td>0.219679</td>
      <td>5.753</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TinyTimeMixerForecaster</td>
      <td>0.229005</td>
      <td>0.706</td>
    </tr>
    <tr>
      <th>1</th>
      <td>MORAI_SMALL</td>
      <td>0.231472</td>
      <td>0.949</td>
    </tr>
    <tr>
      <th>9</th>
      <td>moirai_large</td>
      <td>0.241710</td>
      <td>6.111</td>
    </tr>
    <tr>
      <th>6</th>
      <td>TinyTimeMixerForecaster</td>
      <td>0.245959</td>
      <td>0.756</td>
    </tr>
    <tr>
      <th>0</th>
      <td>MORAI_LARGE</td>
      <td>0.250513</td>
      <td>5.764</td>
    </tr>
    <tr>
      <th>2</th>
      <td>MORAI_SMALL</td>
      <td>0.268142</td>
      <td>0.881</td>
    </tr>
    <tr>
      <th>7</th>
      <td>TinyTimeMixerForecaster</td>
      <td>0.279474</td>
      <td>0.759</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TTM</td>
      <td>0.280884</td>
      <td>0.831</td>
    </tr>
    <tr>
      <th>11</th>
      <td>MORAI_SMALL</td>
      <td>0.294290</td>
      <td>1.442</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TTM</td>
      <td>0.299915</td>
      <td>0.806</td>
    </tr>
    <tr>
      <th>12</th>
      <td>MORAI_LARGE</td>
      <td>0.316691</td>
      <td>5.991</td>
    </tr>
    <tr>
      <th>10</th>
      <td>TTM</td>
      <td>0.356591</td>
      <td>0.906</td>
    </tr>
  </tbody>
</table>
</div>

Another potential bug replicated here is the multirow output per model described here https://github.com/sktime/sktime/issues/8563.

**Versions**
```python
System:
    python: 3.12.10 (main, May 30 2025, 05:39:07) [MSC v.1943 64 bit (AMD64)]
executable: [c:\Users\X\projects\benchmarking-fm\.venv\Scripts\python.exe](file:///C:/Users/X/projects/benchmarking-fm/.venv/Scripts/python.exe)
   machine: Windows-11-10.0.22631-SP0

Python dependencies:
          pip: None
       sktime: 0.38.3
      sklearn: 1.7.0
       skbase: 0.12.3
        numpy: 2.1.3
        scipy: 1.15.3
       pandas: 2.3.1
   matplotlib: 3.10.3
       joblib: 1.5.1
        numba: 0.61.2
  statsmodels: 0.14.5
     pmdarima: None
statsforecast: 2.0.2
      tsfresh: None
      tslearn: None
        torch: 2.7.1
   tensorflow: None
```

### PR DESCRIPTION
**Pull request by jgyasu**
Fixes #8569 

(Stacked on top of #8666, please merge in that order)

As reported in the issue above, the benchmarking result file (either CSV or JSON) contained duplicated results. For example, if from a previous run, the result file contained a model-task combination, it was still getting added to the file if the same model-task combination was present in a new benchmark run, whether after restarting the kernel or without it. 

It is expected that the benchmarks are not run again for an exiting model-task pair and even if it is forceefully run with `force_rerun`, it should still contain only one result.

This PR fixes the above bug.

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks like a fix! But the devil is in the details. Can we test it somehow? The general rationale is that each bugfix should be accompanied by a test that would fail prior but passes after.
Naively thinking, testing should be easy hopefully?

(also, the PR name should start with [BUG] since it is a bugfix)

**Review by fkiraly — CHANGES_REQUESTED** [review]
Could you add a test please?

**Comment by jgyasu** [comment]
Yes, I will add the tests after the break, I am traveling rn 🙂

**Review by fkiraly — APPROVED** [review]
Thanks!

Btw, I notice the duplication between `classification.py` and `forecasting.py` - this is an indicator for "refactorability", where general utility code should probably get pushed up to an abstract parent class.

This is out of the scope of this PR though - I have opened an issue here https://github.com/sktime/sktime/issues/8661
(which you may or may not like to pick up)

**Comment by jgyasu** [comment]
Please do not merge this yet (or unapprove it if possible). I noticed something and I want to confirm if I am thinking correctly.

**Review by fkiraly — CHANGES_REQUESTED** [review]
ok - if there is a bug in the current implementation, or if you are unsure about an edge case, then adding a test for that as well would be appreciated.

File: sktime/benchmarking/forecasting.py [hunk 1]
```diff
@@ -388,17 +388,19 @@ def _run(self, results_path: str, force_rerun: Union[str, list[str]] = "none"):
 
         for task_id, task in self.tasks.entities.items():
             for estimator_id, estimator in self.estimators.entities.items():
-                if results.contains(task_id, estimator_id) and (
```
Comment (fkiraly) [inline sktime/benchmarking/forecasting.py]: is this an accidental revert? `forece_rerun="none"` is no longer handled?

Comment (fkiraly) [reply sktime/benchmarking/forecasting.py]: or is it in the last elif, combined with input assumptions?

File: sktime/forecasting/patch_tst.py [hunk 2]
```diff
@@ -1,670 +0,0 @@
-# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
```
Comment (fkiraly) [inline sktime/forecasting/patch_tst.py]: accidental delete, please do not delete

**Review by fkiraly — CHANGES_REQUESTED** [review]
I merged the changes from #8666.

Repeating my question here: is the `"none"` case properly treated?

**Comment by jgyasu** [comment]
> is the `"none"` case properly treated?

Probably no, I am working on the fix. I had the dataset downloader PR on the first priority.

**Comment by jgyasu** [comment]
This has a deeper issue. The code producing the error provided by Manuel in the issue #8569 uses `force_rerun=True` and I was trying to fix the bug and using that example to check, and I thought I found a fix. But here's the catch, `force_rerun` does not accept boolean values and has no logic to handle it, so `force_rerun=True` is wrong, it should be `force_rerun="all"`, `force_rerun=some_list`, or `force_rerun="none". Using these still produces the same bug after my "fix". 

Now is a good time to rethink the result storage and resumability mechanism from scratch, I will note my thoughts down and get it reviewed.

**Comment by fkiraly** [comment]
are you using "test driven development"? I feel you might not, and that could be the reason that you might feel turning in circles.

I strongly recommend this, here is the recipe:

1. write down code vignettes that you know must work. The more representative examples the better.
2. write down, first on a piece of paper, behaviour that this must exhibit, e.g., outputs returned, files created, content of files, etc.
3. combine the above as code plus asserts in a test. Add these tests to your branch.
4. at the start, it will be failing, but now you develop against the tests directly, so you increasingly get closer to the desired behaviour.

It is crucial to do the above in this exact sequence and not the other way round, e.g., trying to fix a complex bug ad-hoc, and then coming up with a test.

In a case like this, behaviour is complex and multi-layered, e.g., what happens to files, what happens to in-memory objects. I do not really see a realistic chance to get it cleanly fixed by another approach.
